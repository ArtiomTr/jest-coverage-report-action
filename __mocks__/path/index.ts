const realPath = jest.requireActual('path');

export const relative = jest.fn(realPath.relative);
