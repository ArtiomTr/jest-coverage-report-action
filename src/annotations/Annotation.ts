import type { operations } from '@octokit/openapi-types';

export type Annotation = Required<
    Required<
        operations['checks/create']['requestBody']['content']['application/json']
    >['output']
>['annotations'][number];
