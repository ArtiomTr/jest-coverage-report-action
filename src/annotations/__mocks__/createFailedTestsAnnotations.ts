export const createFailedTestsAnnotations = jest.fn();
