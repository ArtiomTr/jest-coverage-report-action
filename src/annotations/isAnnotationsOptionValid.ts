const validChoices = ['all', 'none'];

export const isAnnotationsOptionValid = (option: string): boolean =>
    validChoices.includes(option);
