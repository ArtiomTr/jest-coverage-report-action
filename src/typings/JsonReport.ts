export type JsonReport = {
    numFailedTestSuites: number;
    numFailedTests: number;
    numPassedTestSuites: number;
    numPassedTests: number;
    numPendingTestSuites: number;
    numPendingTests: number;
    numRuntimeErrorTestSuites: number;
    numTodoTests: number;
    numTotalTestSuites: number;
    numTotalTests: number;
    openHandles?: unknown[];
    snapshot: Snapshot;
    startTime: number;
    success: boolean;
    testResults?: TestResult[];
    wasInterrupted: boolean;
    coverageMap: CoverageMap;
};

export type Snapshot = {
    added: number;
    didUpdate: boolean;
    failure: boolean;
    filesAdded: number;
    filesRemoved: number;
    filesRemovedList?: unknown[];
    filesUnmatched: number;
    filesUpdated: number;
    matched: number;
    total: number;
    unchecked: number;
    uncheckedKeysByFile?: unknown[];
    unmatched: number;
    updated: number;
};

export type TestResult = {
    assertionResults?: AssertionResult[];
    endTime: number;
    message: string;
    name: string;
    startTime: number;
    status: string;
    summary: string;
};

export type AssertionResult = {
    ancestorTitles?: string[];
    failureMessages?: string[];
    fullName: string;
    location: Location;
    status: string;
    title: string;
};

export type Location = {
    column?: number;
    line: number;
};

export type Range = {
    start?: Location;
    end?: Location;
};

export type CoverageMap = Record<string, FileCoverage>;

export type FileCoverage = {
    path: string;
    statementMap: StatementMap;
    fnMap: FunctionMap;
    branchMap: BranchMap;
    s: HitMap;
    f: HitMap;
    b: ArrayHitMap;
};

export type StatementMap = Record<number, StatementCoverage>;

export type StatementCoverage = {
    start: Location;
    end: Location;
};

export type FunctionMap = Record<number, FunctionCoverage>;

export type FunctionCoverage = {
    name: string;
    decl: Range;
    loc: Range;
};

export type BranchMap = Record<number, BranchCoverage>;

export type BranchCoverage = {
    loc: Range;
    type: string;
    locations?: Range[];
};

export type HitMap = Record<number, number>;

export type ArrayHitMap = Record<number, number[]>;
