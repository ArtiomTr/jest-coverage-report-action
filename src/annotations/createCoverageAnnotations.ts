import { relative } from 'path';

import { Annotation } from './Annotation';
import { JsonReport, Location } from '../typings/JsonReport';
import { i18n } from '../utils/i18n';
import { isValidNumber } from '../utils/isValidNumber';

const getLocation = (
    start: Location = { line: 0 },
    end: Location = { line: 0 }
): {
    start_line?: number;
    end_line?: number;
    start_column?: number;
    end_column?: number;
} => ({
    start_line: start.line,
    end_line: end.line,
    start_column:
        start.line === end.line && start.column !== null && end.column !== null
            ? start.column
            : undefined,
    end_column:
        start.line === end.line && start.column !== null && end.column !== null
            ? end.column
            : undefined,
});

export const createCoverageAnnotations = (
    jsonReport: JsonReport
): Array<Annotation> => {
    const annotations: Partial<Annotation>[] = [];

    Object.entries(jsonReport.coverageMap).forEach(
        ([fileName, fileCoverage]) => {
            const normalizedFilename = relative(process.cwd(), fileName);
            Object.entries(fileCoverage.statementMap).forEach(
                ([statementIndex, statementCoverage]) => {
                    if (fileCoverage.s[+statementIndex] === 0) {
                        annotations.push({
                            ...getLocation(
                                statementCoverage.start,
                                statementCoverage.end
                            ),
                            path: normalizedFilename,
                            annotation_level: 'warning',
                            title: i18n('notCoveredStatementTitle'),
                            message: i18n('notCoveredStatementMessage'),
                        });
                    }
                }
            );

            Object.entries(fileCoverage.branchMap).forEach(
                ([branchIndex, branchCoverage]) => {
                    if (branchCoverage.locations) {
                        branchCoverage.locations.forEach(
                            (location, locationIndex) => {
                                if (
                                    fileCoverage.b[+branchIndex][
                                        locationIndex
                                    ] === 0
                                ) {
                                    annotations.push({
                                        ...getLocation(
                                            location.start,
                                            location.end
                                        ),
                                        path: normalizedFilename,
                                        annotation_level: 'warning',
                                        title: i18n('notCoveredBranchTitle'),
                                        message: i18n(
                                            'notCoveredBranchMessage'
                                        ),
                                    });
                                }
                            }
                        );
                    }
                }
            );

            Object.entries(fileCoverage.fnMap).forEach(
                ([functionIndex, functionCoverage]) => {
                    if (fileCoverage.f[+functionIndex] === 0) {
                        annotations.push({
                            ...getLocation(
                                functionCoverage.decl.start,
                                functionCoverage.decl.end
                            ),
                            path: normalizedFilename,
                            annotation_level: 'warning',
                            title: i18n('notCoveredFunctionTitle'),
                            message: i18n('notCoveredFunctionMessage'),
                        });
                    }
                }
            );
        }
    );

    return annotations.filter(
        (annotation): annotation is Annotation =>
            isValidNumber(annotation.start_line) &&
            isValidNumber(annotation.end_line)
    );
};
