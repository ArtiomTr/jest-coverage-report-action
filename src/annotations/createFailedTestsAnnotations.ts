import { relative } from 'path';

import stripAnsi from 'strip-ansi';

import { Annotation } from './Annotation';
import { JsonReport } from '../typings/JsonReport';
import { isValidNumber } from '../utils/isValidNumber';

export const createFailedTestsAnnotations = (
    jsonReport: JsonReport
): Array<Annotation> => {
    const testResults = jsonReport.testResults;

    if (!testResults) {
        return [];
    }

    const annotations: Array<Partial<Annotation>> = [];

    const cwd = process.cwd();

    testResults.forEach(({ assertionResults, name: testResultFilename }) => {
        if (!assertionResults) {
            return;
        }

        annotations.push(
            ...assertionResults
                .filter(({ status }) => status === 'failed')
                .map<Annotation>(
                    ({ location, ancestorTitles, title, failureMessages }) => ({
                        annotation_level: 'failure',
                        path: relative(cwd, testResultFilename),
                        start_line: location?.line,
                        end_line: location?.line,
                        title: ancestorTitles?.concat(title).join(' > '),
                        message: stripAnsi(failureMessages?.join('\n\n') ?? ''),
                    })
                )
        );
    });

    return annotations.filter(
        (value): value is Annotation =>
            isValidNumber(value.start_line) && isValidNumber(value.end_line)
    );
};
