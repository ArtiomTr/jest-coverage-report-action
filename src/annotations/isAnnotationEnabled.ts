import { AnnotationType } from '../typings/Options';

export const isAnnotationEnabled = (
    option: string,
    annotationName: AnnotationType
): boolean => option === 'all' || option === annotationName;
