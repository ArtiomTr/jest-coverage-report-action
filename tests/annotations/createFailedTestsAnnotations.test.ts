import { relative } from 'path';

import { createFailedTestsAnnotations } from '../../src/annotations/createFailedTestsAnnotations';
import { JsonReport } from '../../src/typings/JsonReport';
import jsonReport from '../mock-data/jsonReport3.json';

jest.mock('path');

describe('createFailedTestsAnnotations', () => {
    it('should create failed tests annotations', () => {
        (relative as jest.Mock<any, any>).mockImplementation(
            (_, second) => second
        );

        expect(
            createFailedTestsAnnotations((jsonReport as unknown) as JsonReport)
        ).toMatchSnapshot();

        (relative as jest.Mock<any, any>).mockClear();
    });

    it('should return empty array', () => {
        expect(createFailedTestsAnnotations({} as JsonReport)).toStrictEqual(
            []
        );
        expect(
            createFailedTestsAnnotations(({
                testResults: [{}],
            } as unknown) as JsonReport)
        ).toStrictEqual([]);
    });
});
