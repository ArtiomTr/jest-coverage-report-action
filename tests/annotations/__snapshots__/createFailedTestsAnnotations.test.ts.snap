// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createFailedTestsAnnotations > should create failed tests annotations 1`] = `
[
  {
    "annotation_level": "failure",
    "end_line": 4,
    "message": "",
    "path": "tests/format/details/getNewFilesCoverage.test.ts",
    "start_line": 4,
    "title": "getNewFilesCoverage > should return new file coverage",
  },
  {
    "annotation_level": "failure",
    "end_line": 29,
    "message": "",
    "path": "tests/format/details/getNewFilesCoverage.test.ts",
    "start_line": 29,
    "title": "getNewFilesCoverage > should return empty object, when base details not specified",
  },
  {
    "annotation_level": "failure",
    "end_line": 23,
    "message": "",
    "path": "tests/format/summary/getTestRunSummary.test.ts",
    "start_line": 23,
    "title": "getTestRunSummary > should return failure summary",
  },
]
`;
